File: app/models/case.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Text, Index, Boolean
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from app.database import Base
import uuid

class Case(Base):
    __tablename__ = "cases"

    id = Column(Integer, primary_key=True, index=True)
    case_id = Column(UUID(as_uuid=True), default=uuid.uuid4, unique=True, index=True)
    title = Column(String(200), index=True)
    description = Column(Text)
    category = Column(String(50), index=True)
    location = Column(String(100))
    status = Column(String(20), default="pending", index=True)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), index=True)
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
    generated_draft = Column(Text)
    applicable_laws = Column(JSONB)
    suggested_ngos = Column(JSONB)
    next_steps = Column(JSONB)

    # Relationships
    user = relationship("User", back_populates="cases")
    feedback = relationship("Feedback", back_populates="case", cascade="all, delete-orphan")
    
    # Add composite indexes for common queries
    __table_args__ = (
        Index('idx_case_user_status', 'user_id', 'status'),
        Index('idx_case_category_status', 'category', 'status'),
        Index('idx_case_created_status', 'created_at', 'status'),
        # GIN index for JSONB fields for efficient querying
        Index('idx_case_applicable_laws_gin', 'applicable_laws', postgresql_using='gin'),
        Index('idx_case_suggested_ngos_gin', 'suggested_ngos', postgresql_using='gin'),
        Index('idx_case_next_steps_gin', 'next_steps', postgresql_using='gin'),
    ) 

File: empower_lex_mobile/lib/models/case.dart
import 'feedback.dart';

class Case {
  final String id;
  final String title;
  final String description;
  final String category;
  final String priority;
  final String status;
  final String generatedDraft;
  final List<Map<String, dynamic>> applicableLaws;
  final List<Map<String, dynamic>> suggestedNGOs;
  final List<String> nextSteps;
  final List<CaseFeedback> feedback;
  final DateTime createdAt;
  final DateTime? updatedAt;

  Case({
    required this.id,
    required this.title,
    required this.description,
    required this.category,
    required this.priority,
    required this.status,
    required this.generatedDraft,
    required this.applicableLaws,
    required this.suggestedNGOs,
    required this.nextSteps,
    required this.feedback,
    required this.createdAt,
    this.updatedAt,
  });

  factory Case.fromJson(Map<String, dynamic> json) {
    return Case(
      id: json['case_id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      category: json['category'] ?? '',
      priority: json['priority'] ?? '',
      status: json['status'] ?? '',
      generatedDraft: json['generated_draft'] ?? '',
      applicableLaws: (json['applicable_laws'] as List<dynamic>?)?.map((law) => 
        Map<String, dynamic>.from(law as Map)).toList() ?? [],
      suggestedNGOs: (json['suggested_ngos'] as List<dynamic>?)?.map((ngo) => 
        Map<String, dynamic>.from(ngo as Map)).toList() ?? [],
      nextSteps: (json['next_steps'] as List<dynamic>?)?.map((step) => 
        step.toString()).toList() ?? [],
      feedback: (json['feedback'] as List<dynamic>?)?.map((f) => 
        CaseFeedback.fromJson(f as Map<String, dynamic>)).toList() ?? [],
      createdAt: DateTime.parse(json['created_at'] ?? DateTime.now().toIso8601String()),
      updatedAt: json['updated_at'] != null ? DateTime.parse(json['updated_at']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'case_id': id,
      'title': title,
      'description': description,
      'category': category,
      'status': status,
      'priority': priority,
      'generated_draft': generatedDraft,
      'applicable_laws': applicableLaws,
      'suggested_ngos': suggestedNGOs,
      'next_steps': nextSteps,
      'feedback': feedback.map((f) => f.toJson()).toList(),
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }
}
